### PR TITLE
Upgrade Thanos to master

### PR DESCRIPTION
**Pull request by codesome**
**What this PR does**:

Upgrades Thanos to master which has upgraded Prometheus

Some dependencies are giving errors and I am on it

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,8 @@
   * `cortex_compactor_block_cleanup_failed_total`
   * `cortex_compactor_block_cleanup_last_successful_run_timestamp_seconds`
 * [ENHANCEMENT] Experimental TSDB: Use shared cache for metadata. This is especially useful when running multiple querier and store-gateway components to reduce number of object store API calls. #2626
+* [ENHANCEMENT] Upgrade Thanos to [f7802edbf830](https://github.com/thanos-io/thanos/commit/f7802edbf830) and Prometheus to [f4dd45609a05](https://github.com/prometheus/prometheus/commit/f4dd45609a05) which is after v2.18.1.
```
Comment (pracucci) [inline CHANGELOG.md:43]: Please add the PR number here

**Comment by pracucci** [comment]
A part from Prometheus changes, I've checked:
- Thanos upgrade: checked diff, LGTM
- github.com/aws/aws-sdk-go v1.29.18 to v1.30.12: checked the [CHANGELOG](https://github.com/aws/aws-sdk-go/blob/master/CHANGELOG.md) and LGTM, but in `v1.30.16` they fixed a bug "Fix S3 client behavior wrt 200 OK response with empty payload" which could be safer if we get in
- github.com/hashicorp/memberlist v0.1.5 to v0.2.0: no changelog available, checked the diff, done some changes in the transport :fingercrossed:
- github.com/prometheus/client_golang v1.5.1 to v1.6.0: checked the [CHANGELOG](https://github.com/prometheus/client_golang/blob/master/CHANGELOG.md) and LGTM
- github.com/golang/protobuf v1.3.3 to v1.4.0: long [release notes for 1.4.0](https://github.com/golang/protobuf/releases/tag/v1.4.0), but no breaking changes affecting this release.
- github.com/armon/go-metrics v0.3.0 to v0.3.3: according to the diff, minor changes
- cloud.google.com/go/storage v1.3.0 to v1.6.0: checked the [release notes](https://github.com/googleapis/google-cloud-go/releases) and LGTM
- cloud.google.com/go/bigtable v1.1.0 to v1.2.0: checked the [release notes](https://github.com/googleapis/google-cloud-go/releases/tag/bigtable%2Fv1.2.0) and LGTM

File: pkg/api/api.go [hunk 2]
```diff
@@ -255,16 +256,17 @@ func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine
 	api := v1.NewAPI(
 		engine,
 		queryable,
-		querier.DummyTargetRetriever{},
-		querier.DummyAlertmanagerRetriever{},
+		func(context.Context) v1.TargetRetriever { return &querier.DummyTargetRetriever{} },
```
Comment (pracucci) [inline pkg/api/api.go:259]: @gotjosh Once this PR is merged, you should be able to work on your metadata API refactoring.

File: pkg/api/api.go [hunk 3]
```diff
@@ -255,16 +256,17 @@ func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine
...
+		func(context.Context) v1.AlertmanagerRetriever { return &querier.DummyAlertmanagerRetriever{} },
 		func() config.Config { return config.Config{} },
 		map[string]string{}, // TODO: include configuration flags
 		v1.GlobalURLOptions{},
 		func(f http.HandlerFunc) http.HandlerFunc { return f },
-		func() v1.TSDBAdmin { return nil }, // Only needed for admin APIs.
-		false,                              // Disable admin APIs.
+		nil,   // Only needed for admin APIs.
+		"",    // TODO(codesome): what should this be?
```
Comment (pracucci) [inline pkg/api/api.go:266]: This is the directory where snapshots, created via the snapshot API, are stored. Snapshot API is disabled if admin API is disabled (we do disable it in the next line), so it's OK passing an empty string, but please add a comment to explain.

File: pkg/api/api.go (hunk 2 above)
Comment (gotjosh) [reply pkg/api/api.go:259]: Thanks for the head's up, pretty excited to see this. We'll also be able to remove all the copy/paste for the alerts/rules API 😄

**Pull request merged**